### PR TITLE
ROX-10730: comment function case mismatch linting

### PR DESCRIPTION
**Pull request by jschnath**
## Description

~Blocked by https://github.com/stackrox/stackrox/pull/1567 which fixes all occurences in our codebase caught by linting it with this linter.~ https://github.com/stackrox/stackrox/pull/1567 has been merged and this branch rebased to master including the fixes

Anytime a declaration has a name whose capitalization doesn't match with the otherwise same name in its comment, this linter will give a warning.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] ~Unit test and regression tests added~
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

## Testing Performed

Manually tested by introducing the type of error this linter should catch and it looks good to me. Also found many occurrences in the codebase, see the PR linked above.

### CODE REVIEW

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	"unicode"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+const doc = `check that there are no functions, interfaces or structs preceded by descriptive comments of mismatching capitalization`
+
+// Analyzer is the analyzer.
```
Comment (vikin91) [inline tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:17]: 6ezxn6

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
+package declarationcommentmismatch
```
Comment (vikin91) [inline tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:1]: I am not saying it is a bad name, but maybe you had an alternative in mind? 😅

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	"go/token"
+	"strings"
+	"unicode"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+const doc = `check that there are no functions, interfaces or structs preceded by descriptive comments of mismatching capitalization`
```
Comment (vikin91) [inline tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:15]: How about: check the capitalization of the first word for a Godoc comment (for a function, interface, struct)

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		return true
+	})
+	return nil, nil
+}
+
+func checkCommentCaseMatches(pass *analysis.Pass, doc *ast.CommentGroup, objectName string, objectType string, position token.Pos) {
+	if doc == nil {
+		return
+	}
+	commentSplit := strings.Split(doc.List[0].Text, " ")
```
Comment (vikin91) [inline tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:88]: How about using `strings.Fields()`?

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	if len(commentSplit) < 2 {
+		return
+	}
+
+	firstObjectLetter := []rune(objectName[0:1])[0]
```
Comment (vikin91) [inline tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:93]: Let's be on the safe side and maybe check the length before accessing the first element.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 4 above)
Comment (vikin91) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:88]: How about checking whether the `doc.List` has at least length of 1?

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+	firstCommentWord := commentSplit[1]
+	firstCommentLetter := []rune(firstCommentWord[0:1])[0]
```
Comment (vikin91) [inline tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:96]: I think there is no way that the first element of `firstCommentWord` will have 0 length, but still it would be possible to add some checks here

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+		if astFunction, ok := n.(*ast.FuncDecl); ok {
+			checkCommentCaseMatches(pass, astFunction.Doc, astFunction.Name.String(), "function", astFunction.Pos())
+			return true
+		}
+
+		if astVar, ok := n.(*ast.ValueSpec); ok {
+			checkCommentCaseMatches(pass, astVar.Doc, astVar.Names[0].String(), "variable", astVar.Pos())
+			return true
+		}
```
Comment (vikin91) [inline tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:56]: How about struct fields - do they belong into this category?

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		if !push {
+			return true
+		}
+
+		if astFile, ok := n.(*ast.File); ok {
+			file := pass.Fset.File(astFile.Pos())
+			if file != nil && strings.HasSuffix(file.Name(), "_test.go") {
+				return false
+			}
+			return true
```
Comment (vikin91) [inline tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:45]: I guess this could be one return statement.

`return file == nil || !strings.HasSuffix(file.Name(), "_test.go")`

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+	firstObjectLetter := []rune(objectName[0:1])[0]
+
+	firstCommentWord := commentSplit[1]
+	firstCommentLetter := []rune(firstCommentWord[0:1])[0]
+
+	if firstCommentWord[1:] == objectName[1:] && ((unicode.IsLower(firstObjectLetter) && unicode.ToUpper(firstObjectLetter) == firstCommentLetter) || (unicode.IsLower(firstCommentLetter) && unicode.ToUpper(firstCommentLetter) == firstObjectLetter)) {
+		pass.Report(analysis.Diagnostic{
+			Pos:     position,
+			Message: fmt.Sprintf("Mismatching capitalization for %s %s and comment starting with %s", objectType, objectName, firstCommentWord),
```
Comment (vikin91) [inline tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:101]: I like to quote the `%s` using single quotes to make them better visible (what if one of them is empty?). But before doing this, please check if we follow one style of doing such things and if yes, then maybe stick to it.

**Review by vikin91 — CHANGES_REQUESTED** [review]
Nice that you added a custom linter - I always wanted to do this, but never found enough apetite and time.

It looks really good, but there are several sharp edges - maybe we could improve them?

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 2 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:1]: I don't know how to express it in a less verbose way that I'm checking whether capitalization between a declaration and a comment mismatches. Any ideas?

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 3 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:15]: I guess that's simpler but it does not explicitly give the information that it compares the capitalization with the function name. Maybe that's okay.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 4 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:88]: I could check the length of `doc.List` instead of checking that it's not `nil` sure, but why is that better?
As for the `strings.Fields()` function I didn't know it existed, looks great for this purpose so I'll switch to using it here.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 5 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:93]: I can do that but technically it's unnecessary since the AST wouldn't parse it as one of the objects I'm filtering for if it didn't have a name of at least length 1.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 7 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:56]: This category is for var and const declarations, structs are handled within `TypeSpec` `StructType`

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 8 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:45]: Good point

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 9 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:101]: Just checked other linters and there the `%s`'s were not quoted so I'll leave them as is. Plus those `%s`'s cannot be empty logibally speaking, or `pass.Report` would never be reached.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 1 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:17]: Blame all the other linters with the exact same comment that I copied 😆

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 2 above)
Comment (lvalerom) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:1]: Maybe just `commentmismatch` I know it's not completely descriptive but I think people will understand it... maybe? 😅

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 4 above)
Comment (vikin91) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:88]: Well, we check the `doc` for nil, but we do not check the `doc.List` for length. if there is any chance that the `List` may be empty, then this what we have now would result in a panic.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 5 above)
Comment (vikin91) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:93]: I do not know the precise behavior of the parser, but I know that there are anonymous functions in go and they might be a corner case. As we do not have any tests for that, I would suggest to be on the safe side.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 6 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:96]: That goes for both of them but sure I'll introduce checks.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 2 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go:1]: Let's call it `commentcapitalization` then

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 1 above)
Comment (vikin91) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: Please be the good guy and write something informative - maybe it will be so good that others will copy it from you?

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 2 above)
Comment (vikin91) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: Well, let's try to approach this systematically.

(1) The comment before the function will be used of godoc generation (at least for the exported functions). For unexported functions, the comment can be anything (even empty). However, if the comment starts with a word that is the same as function name, we want to make sure that the capitalization matches.

(2) From this description, I am extracting keywords: comment godoc capitalization match.

(3) From the keywords, I try to build something pronounceable: `godoc-comment-capitalization-match`. This could be shortened maybe, so `godoc-capitalization-match`. If the voice must be negative (match vs. mismatch), then maybe `godoc-capitalization-mismatch`. Now just remove the `-` signs, and we have a name.

Feel free to apply this by yourself here. Maybe you will find different keywords?

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 9 above)
Comment (vikin91) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: This is fine not to quote this.

I am having a second though here. This message will be displayed to the user, so maybe it should include some more information. I am thinking of:

1. Telling the user that this has happened because the first word of the comment is the same as the function* name
2. Hinting how to fix that - we will not see any problems where the comment needs to be changed from `// foo` into `// Foo` - this will be handled by govet. So all cases that we discover are about changing `// Foo` into `// foo`. Maybe we can say it directly? Example: _This comment should start with non-capital rune (<explanation why>)_

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 1 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: I mean `Analyzer` is a `analysis.Analyzer` from the `analysis` package of the Go standard lib. It analyzes code, doesn't the name itself define it plenty? I could write `Analyser is responsible for analyzing code` and your meme still holds true.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 4 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: That's fair, better to check than trust it too much.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 9 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: Good idea. I've done it for both cases; even if govet already covers half it can't hurt. I've now also changed the position in the file shown to the user from the position of the declaration to the position of the comment. I think that makes more sense now that I directly recommend what change to make.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 1 above)
Comment (Maddosaurus) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: > Analyzer is a analysis.Analyzer from the analysis package of the Go standard lib. [It analyzes code]

I mean, I would love to have this as the code comment instead of the current one :D
I immediately know way more about it than reading the current comment.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 9 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: Now it only checks for cases where the function is not capitalized

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 2 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: This seems fine, I'll just go with that name then since there's been enough back and forth.

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 1 above)
Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: Sure lol

File: tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	})
+	return nil, nil
+}
+
+func checkCommentCaseMatches(pass *analysis.Pass, doc *ast.CommentGroup, objectName string, objectType string, position token.Pos) {
+	if len(doc.List) < 1 {
+		return
+	}
+	commentSplit := strings.Fields(doc.List[0].Text)
+	if len(commentSplit) < 2 {
```
Comment (vikin91) [inline tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go:94]: Why <2 and not <1?

File: tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+		return
+	}
+	if len(objectName) < 1 {
+		return
+	}
```
Comment (vikin91) [inline tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go:99]: Let's maybe connect these two ifs into one using `||` - would that be okay?

File: tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+	if len(firstCommentWord) < 1 {
+		return
+	}
+	firstCommentLetter := []rune(firstCommentWord[0:1])[0]
+
+	if firstCommentWord[1:] == objectName[1:] {
+		if unicode.IsLower(firstObjectLetter) && unicode.ToUpper(firstObjectLetter) == firstCommentLetter {
+			pass.Report(analysis.Diagnostic{
+				Pos:     position,
+				Message: fmt.Sprintf("Mismatching capitalization for %s %s and comment starting with %s, please change the comment to \"// %s\"", objectType, objectName, firstCommentWord, objectName),
```
Comment (vikin91) [inline tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go:112]: Should the last `%s` be the `objectName`? I guess, we wanted to present there the corrected form of the comment, right? 
Could we maybe use something like: `change the comment to start with: // %s` and then use the `strings.ToLower(firstCommentWord[0])+firstCommentWord[1:]` as the object (I haven't tested it, just trying to codify the idea)?

Comment (jschnath) [reply tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go:112]: That's actually what I had written it as first, then changed to this since the code is simpler. Can revert the code I guess.

File: tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go (hunk 10 above)
Comment (jschnath) [reply tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go:94]: The first element split out of the comment is "//" so I'm looking for the second

File: tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go (hunk 1 above)
Comment (vikin91) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: A good comment would tell us why the thing is here (and optionally what the thing does). So far, I know what it does, but I can only suspect why it is here as a global variable. Guess: I think that each package must have an exported object called Analyzer. Is this correct?

Comment (jschnath) [reply tools/roxvet/analyzers/declarationcommentmismatch/analyzer.go]: I think that is correct because all other linters have one too.

File: tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+	if len(firstCommentWord) < 1 {
+		return
+	}
+	firstCommentLetter := []rune(firstCommentWord[0:1])[0]
+
+	if firstCommentWord[1:] == objectName[1:] {
+		if unicode.IsLower(firstObjectLetter) && unicode.ToUpper(firstObjectLetter) == firstCommentLetter {
+			pass.Report(analysis.Diagnostic{
+				Pos:     position,
+				Message: fmt.Sprintf("Mismatching capitalization for %s %s and comment starting with %s\n please change the comment to start with \"// %d%s\"", objectType, objectName, firstCommentWord, unicode.ToLower(firstCommentLetter), firstCommentWord[1:]),
```
[the remark this answers is not in the document]
Comment (jschnath) [reply tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go:109]: Very good catch, I didn't think to test the new output message. I'll fix it.

[the remark this answers is not in the document]
Comment (jschnath) [reply tools/roxvet/analyzers/godoccapitalizationmismatch/analyzer.go]: Fixed and manually tested it too this time.

**Review by vikin91 — APPROVED** [review]
Thanks for improving and fixing!

(I added a word to the message that I missed in my suggestion)
